Implement collections

tripledoublev opened the pull request:
collections uses cbord

tripledoublev committed:
Add create_collection and test_create_collection

tripledoublev committed:
Add collection related methods and include new test

tripledoublev committed:
Ensure the path is absolute

tripledoublev committed:
Add debug print statements for file and collection operations

RangerMauve commented on src/iroh.rs:
```diff
@@ -380,6 +411,180 @@ impl VeilidIrohBlobs {
 
         return Ok(read);
     }
+    pub async fn create_collection(&self, collection_name: String) -> anyhow::Result<Hash> {
```
```suggestion
    pub async fn create_collection(&self, collection_name: &String) -> anyhow::Result<Hash> {
```

This and other params should probably be borrows since we don't mutate them or pass them off anywhere else

tripledoublev replied:
b242bac685d08ad7d41afe3a2d737029b4341674

tripledoublev committed:
Update serde and serde_cbor versions

tripledoublev committed:
Refactor collections to use cbord

tripledoublev committed:
Use upload from path instead of upload from stream

RangerMauve commented on src/iroh.rs:
```diff
@@ -60,14 +77,39 @@ impl VeilidIrohBlobs {
+        );
+    
+        // Load the collection_hashes HashMap
+        let hash_file_path = base_dir.join("collection_hashes_hash");
```
We shouldn't track collection hashes like this. It should just be store tags so we can avoid needing to have mutextes and stuff on the veilid iroh blobs class

RangerMauve commented on src/iroh.rs:
```diff
@@ -380,6 +474,161 @@ impl VeilidIrohBlobs {
+        println!("Creating a new empty collection with name: {}", collection_name);
+        
+        // Create a new empty collection
+        let collection = Collection::default();
```
Instead of Collection we should use HashMap and save/load that hash map as CBOR like it was a regular file

RangerMauve reviewed (changes requested):
Other than the Collection thing this is looking good. Apologies I wasn't more clear about what I meant regarding the hashmap!

tripledoublev committed:
Refactor create_collection, set_file, get_file, delete_file, and list_files functions for better store interaction.
- Refactored collection hash retrieval to use store tags directly.
- Improved error handling and logging in file operations.
- Added new test cases for missing collections, file overwrites, and concurrent operations.

tripledoublev committed:
Refactor method signatures to use borrowed parameters

RangerMauve commented on src/iroh.rs:
```diff
@@ -474,159 +422,281 @@ impl VeilidIrohBlobs {
-        println!("Uploading the CBOR data from file to the store...");
-
-        // Upload the CBOR data to the store using upload_from_path
         let collection_hash = self.upload_from_path(absolute_temp_path).await?;
```
Before merging we should make sure we upload from stream instead

tripledoublev replied:
Switched to upload_from_stream as part of the encapsulated method -> `update_collection`

tripledoublev replied:
b3fbc4ddf3cab612a93f5f9eef78632ab131cd99

RangerMauve commented on src/iroh.rs:
```diff
@@ -474,159 +422,281 @@ impl VeilidIrohBlobs {
+        println!("Attempting to retrieve tag for collection: {}", collection_name);
+
+        // Retrieve the collection hash using `collection_hash` (use this to get the collection)
+        let collection_hash = match self.collection_hash(&collection_name).await {
```
Use `?` here for readability, we can use whatever error `self.collection_hash` returns as is

RangerMauve commented on src/iroh.rs:
```diff
@@ -474,159 +422,281 @@ impl VeilidIrohBlobs {
+
+        // Fetch collection data by its hash
+        let collection_entry = self.store.get(&collection_hash).await?;
+        let mut collection: HashMap<String, Hash> = if let Some(entry) = collection_entry {
```
Is `.get().await?.some_or_else(||anyhow)?` instead of the if let

tripledoublev replied:
deb3095d8352c1eb85155d79479a7a9f3eb78345

RangerMauve commented on src/iroh.rs:
```diff
@@ -474,159 +422,281 @@ impl VeilidIrohBlobs {
+        // Fetch collection data by its hash
+        let collection_entry = self.store.get(&collection_hash).await?;
+        let mut collection: HashMap<String, Hash> = if let Some(entry) = collection_entry {
+            let mut reader = entry.data_reader();
```
Would using `read_bytes` be better here? it can handle the data reader stuff via reading the stream

RangerMauve commented on src/iroh.rs:
```diff
@@ -474,159 +422,281 @@ impl VeilidIrohBlobs {
+        println!("New collection hash: {:?}", new_collection_hash);
+
+        // Store the new collection hash with the tag
+        self.store_tag(&collection_name, &new_collection_hash).await?;
```
We should encapsulate the logic of encoding/hashing/tagging into a function

e.g. `let hash = self.update_collection(name, collection).await?`

tripledoublev replied:
b3fbc4ddf3cab612a93f5f9eef78632ab131cd99

RangerMauve commented on src/iroh.rs:
```diff
@@ -474,159 +422,281 @@ impl VeilidIrohBlobs {
-            let mut map = self.collection_hashes.write().await;
-            map.insert(collection_name, new_collection_hash);
+        // Introduce a small delay to ensure the tag is properly stored and synced
+        tokio::time::sleep(Duration::from_millis(100)).await;
```
Is there an error if you remove this? Storing the tag should resolve once it's fully finished, no?

tripledoublev replied:
No errors. It was unnecessary delay. c50310e8400aca6c42cf6e2d481beca9dc04b03f

RangerMauve commented on src/iroh.rs:
```diff
@@ -474,159 +422,281 @@ impl VeilidIrohBlobs {
+
+
+        // Verify the tag immediately after storing
+        let verified_hash = self.get_tag(&collection_name).await?;
```
Did you have errors around this in your tests? It sounds like a bug if the tag isn't persisting. We should fix it and remove this verification code

tripledoublev replied:
Unnecessary verification. Removed it c50310e8400aca6c42cf6e2d481beca9dc04b03f

RangerMauve commented on src/iroh.rs:
```diff
@@ -474,159 +422,281 @@ impl VeilidIrohBlobs {
     }
+    pub async fn get_file(&self, collection_name: &String, path: &String) -> Result<Hash> {
+        // Retrieve the current collection hash from the tag
+        let collection_hash = self.get_tag(&collection_name).await?;
```
We should have a method which gets the collection to deduplicate effort. like `let entry = self.get_collection(name)` which will get the hash from the tag and decode the cbor.

RangerMauve commented on src/iroh.rs:
```diff
@@ -474,159 +422,281 @@ impl VeilidIrohBlobs {
+
+                    // Extract the bytes and convert them to a string
+                    let tag_bytes = tag.borrow();
+                    let tag_str = std::str::from_utf8(tag_bytes).unwrap_or("");
```
Instead lets conver the name to byes and do a `tag_bytes.eq` to save on parsing

tripledoublev replied:
I updated get_tag to compare bytes to save on parsing
dc0b4b490faaf22a16bf5d539afa5393be33e901

RangerMauve commented on src/iroh.rs:
```diff
@@ -474,159 +422,281 @@ impl VeilidIrohBlobs {
+                        println!("Tag {} does not match {}", tag_str, collection_name);
+                    }
+                }
+                Err(e) => {
```
Prefer to use `tag_result?` or `.map_err` to these match statements

tripledoublev replied:
Switched to using the ? operator and map_err
ff27d38067a00229ac9fb03d159bc4ade2b228aa

RangerMauve commented on src/iroh.rs:
```diff
@@ -474,159 +422,281 @@ impl VeilidIrohBlobs {
+        Err(anyhow!("Tag not found for collection: {}", collection_name))
+    }
+    
+    pub async fn list_tags(&self) -> Result<()> {
```
Should this be removed?

tripledoublev replied:
Yes! I removed list_tags
89e8c05958dc19c7827d46d8034ecd872eee3a9f

RangerMauve reviewed (changes requested):
Good work, mostly formatting comments. Feel free to merge after addressing, I can also try to help with the stream upload of the collection bytes tonight

tripledoublev committed:
Remove unnecessary verification and delay

tripledoublev committed:
Update get_tag for direct bytes comparison

tripledoublev committed:
Remove list_tags

tripledoublev committed:
Refactor get_tag to use ? operator and map_err for error handling

tripledoublev committed:
Refactor set_file to use ?, ok_or_else, and map_err

tripledoublev committed:
Refactor collection update logic into encapsulated `update_collection` method
